### PR TITLE
Setup: fix #37755, add component.factory to DIC

### PR DESCRIPTION
**Pull request by nhaagen**
https://mantis.ilias.de/view.php?id=37755

### CODE REVIEW

**Comment by mjansenDatabay** [comment]
@nhaagen I pushed a commit to your PR branch where I applied the same change for two further objectives.

**Comment by mjansenDatabay** [comment]
@klees Maybe we should merge (if there aren't any objections) this before 8.5 will be released :-)?

**Comment by klees** [comment]
Sry people, didn't have the time to look into this until now and won't sacrifice correctness for speed. I informed @fwolf-ilias and will probably have the time to merge this for 8.5.

**Comment by klees** [comment]
Hi @mjansenDatabay,

do you have according reports for the changes you have added? I really wonder why we did not notice during testing and what has changed in the meantime. From my perspective, the thingies in that "abomination" part in the `Objectives` should become less, not more. Why do we need these additions now?

Kind regards!

**Comment by mjansenDatabay** [comment]
Hi @klees ,

> I [...] won't sacrifice correctness for speed

+1

> do you have according reports for the changes you have added

I added a stack trace (see: Mantis) for one of the changed objectives.

>  I really wonder why we did not notice during testing and what has changed in the meantime.

To be honest, I think plugins have not been tested seriously, neither by developers, nor by QA. I can name multiple plugin slots in ILIAS 8 where I can guarantee no one ever executed a test for.
But I agree, as far as I remember the repository object slot did not cause these issues in the early ILIAS 8 releases.

The error always occurs if there is at least one repository object plugin and an instance of "ilObjectDefinition" is created in one of the objectives.

Best regards,
Michael

File: Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php [hunk 1]
```diff
@@ -242,6 +243,7 @@ public function raise($a_component, $a_event, $a_parameter = ""): void
         };
         $GLOBALS["DIC"]["ilObjDataCache"] = new ilObjectDataCache();
         $GLOBALS["DIC"]["ilSetting"] = new ilSetting();
+        $GLOBALS["DIC"]["component.factory"] = $component_factory;
```
Comment (klees) [inline Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php]: @mjansenDatabay Would you be so kind and add that `component.repository` here as well?

**Comment by klees** [comment]
FYI: I wondered why we never found these problems during testing, and I guess I have found the explanations I was looking for, and one suggestion for a process-related improvement as well:

https://mantis.ilias.de/view.php?id=37755#c95896

**Comment by fwolf-ilias** [comment]
@klees I'm going to postpone the release of 8.5 until tomorrow. Can I have an estimate whether this PR will be merged by tomorrow sometime?

File: Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php (hunk 1 above)
Comment (klees) [reply Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php]: On it =)

Comment (mjansenDatabay) [reply Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php]: I can change this, just ping me in Discord :-).

-- Update --

OK: https://github.com/ILIAS-eLearning/ILIAS/pull/6277/commits/7c4b299edc68a1160042ab22f149d17176baffe4

Comment (klees) [reply Services/Component/classes/Setup/class.ilComponentUpdatePluginObjective.php]: Found some more oddities here...

**Comment by klees** [comment]
Thx y'all!

**Comment by klees** [comment]
Ping @fwolf-ilias: Thanks for waiting for this!

**Comment by klees** [comment]
Picked to trunk.

**Pull request merged**